src/main/scala/mesosphere/marathon/MarathonSchedulerActor.scala

```diff
@@ -163,127 +161,166 @@ class MarathonSchedulerActor(
-      val res = deploy(driver, plan)
-      if (origSender != Actor.noSender) origSender ! cmd.answer
-      res
+    val res = withLockFor(ids) {
+      deploy(driver, plan)
+    }
+
+    res match {
+      case Success(_) =>
+        if (origSender != Actor.noSender) origSender ! cmd.answer
+      case Failure(e: LockingFailedException) if cmd.force =>
+        deploymentManager ! CancelConflictingDeployments(plan)
+        val cancellationHandler = context.system.scheduler.scheduleOnce(
+          cancellationTimeout,
+          self,
+          CancellationTimeoutExceeded)
+
+        context.become(awaitCancellation(plan, origSender, cancellationHandler))
+      case Failure(e: LockingFailedException) =>
+        deploymentManager.ask(RetrieveRunningDeployments)(2.seconds)
+          .mapTo[RunningDeployments]
+          .foreach {
+            case RunningDeployments(plans) =>
+              val relatedDeploymentIds: Seq[String] = plans.collect {
+                case (p, _) if p.affectedApplicationIds.intersect(plan.affectedApplicationIds).nonEmpty => p.id
+              }
+              origSender ! CommandFailed(cmd, AppLockedException(relatedDeploymentIds))
+          }
     }
   }
 
-  def deploy(driver: SchedulerDriver, plan: DeploymentPlan): Future[Unit] = {
-    deploymentRepository.store(plan).flatMap { _ =>
-      val promise = Promise[Any]()
-      val promiseActor = context.actorOf(Props(classOf[PromiseActor], promise))
-      val msg = PerformDeployment(driver, plan)
-      deploymentManager.tell(msg, promiseActor)
+  def deploy(driver: SchedulerDriver, plan: DeploymentPlan): Unit = {
+    deploymentRepository.store(plan).foreach { _ =>
+      deploymentManager ! PerformDeployment(driver, plan)
+    }
+  }
 
-      val res = promise.future.map(_ => ())
+  def deploymentSuccess(plan: DeploymentPlan): Unit = {
+    log.info(s"Deployment of ${plan.target.id} successful")
+    eventBus.publish(DeploymentSuccess(plan.id))
+    deploymentRepository.expunge(plan.id)
+  }
 
-      res andThen {
-        case Success(_) =>
-          log.info(s"Deployment of ${plan.target.id} successful")
-          eventBus.publish(DeploymentSuccess(plan.id))
-          deploymentRepository.expunge(plan.id)
-
-        case Failure(e) =>
-          log.error(e, s"Deployment of ${plan.target.id} failed")
-          plan.affectedApplicationIds.foreach(appId => taskQueue.purge(appId))
-          eventBus.publish(DeploymentFailed(plan.id))
-          if (e.isInstanceOf[DeploymentCanceledException])
-            deploymentRepository.expunge(plan.id)
-      }
-    }
+  def deploymentFailed(plan: DeploymentPlan, reason: Throwable): Unit = {
+    log.error(reason, s"Deployment of ${plan.target.id} failed")
+    plan.affectedApplicationIds.foreach(appId => taskQueue.purge(appId))
+    eventBus.publish(DeploymentFailed(plan.id))
+    if (reason.isInstanceOf[DeploymentCanceledException])
+      deploymentRepository.expunge(plan.id)
```

Comment:
I guess the main concern is usability.  From the user's perspective, if the deployment goes away then users have no way to tell whether it completed or failed.  Maybe it would be good to keep a bounded buffer of deployments reaching into the past (just mark them completed).  The REST API for LXD allows users to hit an endpoint to get streaming events for the progress of a background task.  It also presents an endpoint for each such process which will block, returning a response only when the process completes.  Something like these could be really useful here.